### DISCUSSION TITLE
Endpoint Tests

### DISCUSSION THREAD

**Discussion by odrya12** [serverpod/serverpod#1376] [Q&A]
There is any built-in api to run endpoint tests(unit)?

The only way I found is to add the client as dev dependency, run the server and then run the tests.

will it be simpler to add dedicated api inside the framework for that?

**Comment by Isakdl** [comment]
Unfortunately, we do not have this yet. It is on our minds and something we plan to build, we need it for testing the Serverpod framework itself as well. 

Today we have solved this as you suggested, by doing end-to-end tests with the client library. This means we spin up the Serverpod server and the database in their own docker containers and then run the test suit against this running instance. There is value in these types of tests too but of course, they are heavier to maintain and run than a unit or integration test. 

For an example of this, you can look at the test project `serverpod_test_server` in the `tests` folder in the serverpod repo.

**Comment by odrya12** [comment]
I think it's important and would be glad to help with that. I might develop something for it. End-to-end tests assess the big picture and could, in some cases, make bug finding harder.

**Reply by Isakdl** [reply]
Yes totally agree! Our plan is to tackle this after the 1.2 release, but if you have some good ideas here feel free to take a stab at it.
